### PR TITLE
fuzz: server config validation fuzz test.

### PR DESCRIPTION
**Pull request by anirudhmurali**
*Title:* fuzz: server config validation fuzz test

*Description:* This fuzzes the bootstrap config message on server validation mode.

*Risk Level:* Low

*Testing:* Tested under oss-fuzz

Signed-off-by: Anirudh M m.anirudh18@gmail.com

### CODE REVIEW

File: test/server/config_validation/BUILD [hunk 1]
```diff
@@ -1,6 +1,17 @@
 licenses(["notice"])  # Apache 2
 
-load("//bazel:envoy_build_system.bzl", "envoy_cc_test", "envoy_package")
+load(
+    "//bazel:envoy_build_system.bzl",
+    "envoy_cc_fuzz_test",
+    "envoy_cc_test",
+    "envoy_cc_test_library",
+    "envoy_package",
+    "envoy_select_hot_restart",
```
Comment (htuch) [inline test/server/config_validation/BUILD]: A number of these are unnecessary; you only added a `envoy_cc_fuzz_test` target, so this should be the only new one needed.

File: test/server/config_validation/BUILD [hunk 2]
```diff
@@ -1,6 +1,17 @@
...
+)
+load(
+    "//source/extensions:all_extensions.bzl",
+    "envoy_all_extensions",
+)
```
Comment (htuch) [inline test/server/config_validation/BUILD]: Unused.

File: test/server/config_validation/config_fuzz_test.cc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "envoy/config/bootstrap/v2/bootstrap.pb.h"
+#include "envoy/config/bootstrap/v2/bootstrap.pb.validate.h"
+
+#include "common/network/address_impl.h"
+#include "common/thread_local/thread_local_impl.h"
+
+#include "server/config_validation/server.h"
+#include "server/drain_manager_impl.h"
+#include "server/hot_restart_nop_impl.h"
```
Comment (htuch) [inline test/server/config_validation/config_fuzz_test.cc]: I think a bunch of these headers aren't needed either; they are vestigial from parts that were removed in `server_fuzz_test`.

File: test/server/config_validation/config_fuzz_test.cc [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#include <fstream>
```
Comment (htuch) [inline test/server/config_validation/config_fuzz_test.cc]: Please add a comment to this file explaining that it is derived from `//test/server:server_fuzz_test.cc` and the substantive difference.

**Comment by htuch** [comment]
@mattklein123

**Comment by mattklein123** [comment]
@anirudhmurali please ping me when this is ready for another pass.

File: test/server/config_validation/BUILD [hunk 5]
```diff
@@ -89,13 +78,12 @@ envoy_cc_fuzz_test(
...
+        "//source/server:options_lib",
         "//source/server:server_lib",
+        "//source/server/config_validation:server_lib",
         "//test/integration:integration_lib",
         "//test/mocks/server:server_mocks",
-        "//test/mocks/stats:stats_mocks",
-        "//source/server:configuration_lib",
-        "//source/server/config_validation:server_lib",
-        "//source/server:options_lib",
         "//test/test_common:environment_lib",
```
Comment (htuch) [inline test/server/config_validation/BUILD]: In general, these deps should reflect the headers in `config_fuzz_test.cc`; can you clean this up?

File: test/server/config_validation/config_fuzz_test.cc [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "envoy/config/bootstrap/v2/bootstrap.pb.h"
+#include "envoy/config/bootstrap/v2/bootstrap.pb.validate.h"
+
+#include "common/network/address_impl.h"
+#include "common/thread_local/thread_local_impl.h"
```
Comment (htuch) [inline test/server/config_validation/config_fuzz_test.cc]: Is this needed below?

File: test/server/config_validation/config_fuzz_test.cc [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "server/config_validation/server.h"
+#include "server/options_impl.h"
+#include "server/server.h"
+#include "server/test_hooks.h"
+
+#include "test/fuzz/fuzz_runner.h"
+#include "test/integration/server.h"
```
Comment (htuch) [inline test/server/config_validation/config_fuzz_test.cc]: Is this needed below?

**Comment by anirudhmurali** [comment]
Running `bazel test //test/server/config_validation:config_fuzz_test -config=clang-tsan --runs_per_test=2000` shows like few (3-7) failing tests out of 2000. All with the same message: `Didn't find a registered implementation for name: 'tls'`.
I tried linking all extensions, and tested, it throws:

`[critical][assert] external/envoy/source/common/thread_local/thread_local_impl.cc:19] assert failure: shutdown_`. Same reason why tsan failed in the previous commits.

The `ASSERT(shutdown_);` statement  in ~InstanceImpl() is what is causing this failure. `shutdown_` seems to be a atomic boolean object. Which part of the code might have set shutdown_ to false?

**Comment by htuch** [comment]
@anirudhmurali I suggest grepping (or using other code search tools) to understand how `shutdown_` is manipulated. I think that adding back the `envoy_all_extensions()` was correct; what needs to be done here is figure out why we race in shutdown. We generally expect code committers to take time to understand the framework and fix bugs like this as a gating step before commit.

How often are you seeing the flake with `shutdown_`?

**Comment by anirudhmurali** [comment]
Used `validateConfig()` instead of creating `ValidationInstance` object. Things work now, all tests pass well, even with `runs_per_test=2000` set. I guess the server crash was because of some improper initialization of `stats_store` or `access_log_lock`.
Should I run more tests to uncover flakes?

File: test/server/config_validation/config_fuzz_test.cc [hunk 8]
```diff
@@ -14,20 +14,18 @@ namespace Server {
 // Derived from //test/server:server_fuzz_test.cc, but starts the server in configuration validation
 // mode (quits upon validation of the given config)
 DEFINE_PROTO_FUZZER(const envoy::config::bootstrap::v2::Bootstrap& input) {
+
```
Comment (htuch) [inline test/server/config_validation/config_fuzz_test.cc]: Nit: Remove this empty line.

**Pull request merged**